File: dotnet/system/workspace/typescript/libs/workspace/domain/json/system/src/lib/database/derivationerror.ts
import { ResponseDerivationError } from '@allors/protocol/json/system';
import { IDerivationError, Role } from '@allors/workspace/domain/system';
import { Session } from '../Session/Session';

export class DerivationError implements IDerivationError {
  constructor(private session: Session, private responseDerivationError: ResponseDerivationError) {}

  get message() {
    return this.responseDerivationError.m;
  }

  get roles(): Role[] {
    return this.responseDerivationError.r.map((r) => {
      return {
        object: this.session.getOne(r[0]),
        relationType: this.session.workspace.metaPopulation.metaObjectByTag.get(r[1]),
      } as Role;
    });
  }
}

File: dotnet/system/workspace/csharp/adapters/allors.workspace.adapters.remote/database/derivationerror.cs
// <copyright file="RemoteDerivationError.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Json.Api;
    using Meta;

    public class DerivationError : IDerivationError
    {
        private readonly ISession session;
        private readonly ResponseDerivationError responseDerivationError;

        public DerivationError(ISession session, ResponseDerivationError responseDerivationError)
        {
            this.session = session;
            this.responseDerivationError = responseDerivationError;
        }

        public string Message => this.responseDerivationError.m;

        public IEnumerable<Role> Roles =>
            from r in this.responseDerivationError.r
            let association = this.session.Instantiate<IObject>(r.i)
            let relationType = (IRelationType)this.session.Workspace.Configuration.MetaPopulation.FindByTag(r.r)
            select new Role(association, relationType);
    }
}
